File: platform/how-to-get-started/2.fetch_httpx_gemini_openai_protocols/2.gemini-openai-streaming.mjs
#!/usr/bin/env node
/**
 * Example: Full Streaming Translation with Live API Calls
 *
 * This example demonstrates:
 * 1. Making a streaming request to Gemini's native API
 * 2. Translating Gemini NDJSON stream to OpenAI SSE format in real-time
 * 3. Aggregating streaming responses into complete messages
 * 4. Handling tool calls in streaming responses
 *
 * This uses the full stream translation functions (async generators).
 */
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { Readable } from 'stream';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const ROOT_DIR = join(__dirname, '../..');

// Import protocol translation
const {
    translateOpenAIRequestToGemini,
    geminiStreamToOpenAI,
    aggregateOpenAIStream,
} = await import(
    join(ROOT_DIR, 'packages_mjs/fetch_undici_gemini_openai_protocols/dist/index.js')
);

// Import fetch client
const { AsyncClient, Timeout, iterLines } = await import(
    join(ROOT_DIR, 'packages_mjs/fetch_undici/dist/index.js')
);

// Import Gemini constants
const {
    GEMINI_ORIGIN,
    GEMINI_CONNECT_TIMEOUT_MS,
    GEMINI_READ_TIMEOUT_MS,
} = await import(
    join(ROOT_DIR, 'packages_mjs/fetch_undici_gemini_openai_constant/dist/index.js')
);

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) {
    console.error('ERROR: GEMINI_API_KEY environment variable not set');
    console.error('Set it with: export GEMINI_API_KEY=your_api_key');
    process.exit(1);
}

// =============================================================================
// Example 1: Simple Streaming Chat
// =============================================================================

console.log('='.repeat(60));
console.log('Example 1: Streaming Chat with Format Translation');
console.log('='.repeat(60));

const openAIRequest = {
    model: 'gemini-2.0-flash',
    messages: [
        { role: 'system', content: 'You are a helpful assistant. Keep responses brief.' },
        { role: 'user', content: 'Write a haiku about programming.' },
    ],
    temperature: 0.7,
    max_tokens: 100,
    stream: true,  // Enable streaming
};

console.log('\nOpenAI-style Request:');
console.log(JSON.stringify(openAIRequest, null, 2));

// Translate to Gemini format
const geminiRequest = translateOpenAIRequestToGemini(openAIRequest);
console.log('\nTranslated to Gemini Request:');
console.log(JSON.stringify(geminiRequest.data, null, 2));

// Make streaming request to Gemini
const GEMINI_STREAM_URL = `${GEMINI_ORIGIN}/v1beta/models/gemini-2.0-flash:streamGenerateContent?alt=sse`;

const client = new AsyncClient({
    timeout: new Timeout({ connect: GEMINI_CONNECT_TIMEOUT_MS, read: GEMINI_READ_TIMEOUT_MS }),
});

try {
    console.log('\n--- Starting Streaming Request ---');
    console.log('URL:', GEMINI_STREAM_URL);

    const response = await client.post(GEMINI_STREAM_URL, {
        headers: {
            'Content-Type': 'application/json',
            'x-goog-api-key': GEMINI_API_KEY,
        },
        json: geminiRequest.data,
    });

    if (!response.isSuccess) {
        const errorText = await response.text();
        console.error('API Error:', response.statusCode, errorText);
        process.exit(1);
    }

    console.log('\n--- Gemini NDJSON Stream (raw) ---');

    // Create an async iterator from the response body
    async function* streamLines(body) {
        for await (const line of iterLines(body)) {
            yield line;
        }
    }

    // Collect lines for translation
    const lines = [];
    for await (const line of streamLines(response.body)) {
        // Gemini with alt=sse sends SSE format
        if (line.startsWith('data: ')) {
            const jsonLine = line.slice(6);
            if (jsonLine.trim()) {
                console.log(`Gemini: ${jsonLine.substring(0, 80)}${jsonLine.length > 80 ? '...' : ''}`);
                lines.push(jsonLine);
            }
        }
    }

    console.log('\n--- Translating to OpenAI SSE Format ---');

    // Create async iterator from collected lines
    async function* iterLines2(lines) {
        for (const line of lines) {
            yield line;
        }
    }

    // Translate the stream
    let fullContent = '';
    for await (const openAISSE of geminiStreamToOpenAI(iterLines2(lines), 'gemini-2.0-flash')) {
        console.log(`OpenAI: ${openAISSE.trim()}`);

        // Parse to extract content
        if (openAISSE.startsWith('data: ') && !openAISSE.includes('[DONE]')) {
            try {
                const chunk = JSON.parse(openAISSE.slice(6));
                const content = chunk.choices?.[0]?.delta?.content || '';
                fullContent += content;
            } catch (e) {
                // Ignore parse errors
            }
        }
    }

    console.log('\n--- Aggregated Response ---');
    console.log('Full content:', fullContent);

} catch (error) {
    console.error('Error:', error.message);
} finally {
    await client.close();
}

// =============================================================================
// Example 2: Streaming with Tool Calls
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 2: Streaming with Tool Calls');
console.log('='.repeat(60));

const toolRequest = {
    model: 'gemini-2.0-flash',
    messages: [
        { role: 'system', content: 'You are a helpful assistant. Use tools when needed.' },
        { role: 'user', content: 'What is the weather in Paris?' },
    ],
    tools: [
        {
            type: 'function',
            function: {
                name: 'get_weather',
                description: 'Get the weather for a location',
                parameters: {
                    type: 'object',
                    properties: {
                        location: { type: 'string', description: 'City name' },
                    },
                    required: ['location'],
                },
            },
        },
    ],
    tool_choice: 'auto',
    stream: true,
};

console.log('\nOpenAI-style Request with Tools:');
console.log(JSON.stringify(toolRequest, null, 2));

// Translate to Gemini
const geminiToolRequest = translateOpenAIRequestToGemini(toolRequest);
console.log('\nTranslated to Gemini:');
console.log(JSON.stringify(geminiToolRequest.data, null, 2));

const client2 = new AsyncClient({
    timeout: new Timeout({ connect: GEMINI_CONNECT_TIMEOUT_MS, read: GEMINI_READ_TIMEOUT_MS }),
});

try {
    const response = await client2.post(GEMINI_STREAM_URL, {
        headers: {
            'Content-Type': 'application/json',
            'x-goog-api-key': GEMINI_API_KEY,
        },
        json: geminiToolRequest.data,
    });

    if (!response.isSuccess) {
        const errorText = await response.text();
        console.error('API Error:', response.statusCode, errorText);
    } else {
        console.log('\n--- Streaming Response ---');

        // Collect and display chunks
        const lines = [];
        for await (const line of iterLines(response.body)) {
            if (line.startsWith('data: ')) {
                const jsonLine = line.slice(6);
                if (jsonLine.trim()) {
                    console.log(`Raw: ${jsonLine.substring(0, 100)}${jsonLine.length > 100 ? '...' : ''}`);
                    lines.push(jsonLine);
                }
            }
        }

        // Parse and show tool calls
        console.log('\n--- Parsed Tool Calls ---');
        for (const line of lines) {
            try {
                const chunk = JSON.parse(line);
                const parts = chunk.candidates?.[0]?.content?.parts || [];
                for (const part of parts) {
                    if (part.functionCall) {
                        console.log(`Tool Call: ${part.functionCall.name}(${JSON.stringify(part.functionCall.args)})`);
                    }
                }
            } catch (e) {
                // Ignore parse errors
            }
        }
    }
} catch (error) {
    console.error('Error:', error.message);
} finally {
    await client2.close();
}

// =============================================================================
// Example 3: Stream Aggregation
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 3: Stream Aggregation Demo');
console.log('='.repeat(60));

// Simulate an OpenAI SSE stream
const simulatedSSELines = [
    'data: {"id":"chatcmpl-1","choices":[{"delta":{"role":"assistant","content":"Hello"}}]}',
    'data: {"id":"chatcmpl-1","choices":[{"delta":{"content":" world"}}]}',
    'data: {"id":"chatcmpl-1","choices":[{"delta":{"content":"!"}}]}',
    'data: {"id":"chatcmpl-1","choices":[{"delta":{},"finish_reason":"stop"}]}',
    'data: [DONE]',
];

console.log('\nSimulated OpenAI SSE Stream:');
for (const line of simulatedSSELines) {
    console.log(line);
}

// Create async iterator
async function* iterSSE(lines) {
    for (const line of lines) {
        yield line;
    }
}

// Aggregate the stream
const aggregated = await aggregateOpenAIStream(iterSSE(simulatedSSELines));

console.log('\nAggregated Result:');
console.log(JSON.stringify(aggregated.data, null, 2));

const content = aggregated.data.choices?.[0]?.delta?.content;
console.log('\nFull Message:', content);

console.log('\n' + '='.repeat(60));
console.log('Streaming Examples Complete');
console.log('='.repeat(60));

File: platform/how-to-get-started/2.fetch_httpx_gemini_openai_protocols/2.gemini-openai-streaming.py
#!/usr/bin/env python3
"""
Example: Full Streaming Translation with Live API Calls

This example demonstrates:
1. Making a streaming request to Gemini's native API
2. Translating Gemini NDJSON stream to OpenAI SSE format in real-time
3. Aggregating streaming responses into complete messages
4. Handling tool calls in streaming responses

This uses the full stream translation functions (async generators).
"""
import os
import sys
import json
import asyncio
from pathlib import Path
from typing import AsyncIterator

# Add packages to path
ROOT_DIR = Path(__file__).parent.parent.parent
sys.path.insert(0, str(ROOT_DIR / "packages_py"))
sys.path.insert(0, str(ROOT_DIR / "packages_py/fetch_httpx_gemini_openai_constant"))

from fetch_httpx import AsyncClient, Timeout
from fetch_httpx_gemini_openai_constant import (
    GEMINI_ORIGIN,
    GEMINI_CONNECT_TIMEOUT_S,
    GEMINI_READ_TIMEOUT_S,
)
from fetch_httpx_gemini_openai_protocols import (
    translate_openai_request_to_gemini,
    gemini_stream_to_openai,
    aggregate_openai_stream,
)

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
if not GEMINI_API_KEY:
    print("ERROR: GEMINI_API_KEY environment variable not set")
    print("Set it with: export GEMINI_API_KEY=your_api_key")
    sys.exit(1)


async def main():
    # =========================================================================
    # Example 1: Simple Streaming Chat
    # =========================================================================

    print("=" * 60)
    print("Example 1: Streaming Chat with Format Translation")
    print("=" * 60)

    openai_request = {
        "model": "gemini-2.0-flash",
        "messages": [
            {"role": "system", "content": "You are a helpful assistant. Keep responses brief."},
            {"role": "user", "content": "Write a haiku about programming."},
        ],
        "temperature": 0.7,
        "max_tokens": 100,
        "stream": True,  # Enable streaming
    }

    print("\nOpenAI-style Request:")
    print(json.dumps(openai_request, indent=2))

    # Translate to Gemini format
    gemini_request = translate_openai_request_to_gemini(openai_request)
    print("\nTranslated to Gemini Request:")
    print(json.dumps(gemini_request.data, indent=2))

    # Make streaming request to Gemini
    GEMINI_STREAM_URL = f"{GEMINI_ORIGIN}/v1beta/models/gemini-2.0-flash:streamGenerateContent?alt=sse"

    async with AsyncClient(timeout=Timeout(connect=GEMINI_CONNECT_TIMEOUT_S, read=GEMINI_READ_TIMEOUT_S)) as client:
        print("\n--- Starting Streaming Request ---")
        print("URL:", GEMINI_STREAM_URL)

        async with client.stream(
            "POST",
            GEMINI_STREAM_URL,
            headers={
                "Content-Type": "application/json",
                "x-goog-api-key": GEMINI_API_KEY,
            },
            json=gemini_request.data,
        ) as response:
            if not response.is_success:
                error_text = await response.aread()
                print(f"API Error: {response.status_code} {error_text.decode()}")
                return

            print("\n--- Gemini SSE Stream (raw) ---")

            # Collect lines for translation
            lines = []
            async for line in response.aiter_lines():
                # Gemini with alt=sse sends SSE format
                if line.startswith("data: "):
                    json_line = line[6:]
                    if json_line.strip():
                        display = json_line[:80] + ("..." if len(json_line) > 80 else "")
                        print(f"Gemini: {display}")
                        lines.append(json_line)

            print("\n--- Translating to OpenAI SSE Format ---")

            # Create async iterator from collected lines
            async def iter_lines(lines_list):
                for line in lines_list:
                    yield line

            # Translate the stream
            full_content = ""
            async for openai_sse in gemini_stream_to_openai(iter_lines(lines), "gemini-2.0-flash"):
                print(f"OpenAI: {openai_sse.strip()}")

                # Parse to extract content
                if openai_sse.startswith("data: ") and "[DONE]" not in openai_sse:
                    try:
                        chunk = json.loads(openai_sse[6:])
                        content = chunk.get("choices", [{}])[0].get("delta", {}).get("content", "")
                        full_content += content
                    except json.JSONDecodeError:
                        pass

            print("\n--- Aggregated Response ---")
            print("Full content:", full_content)

    # =========================================================================
    # Example 2: Streaming with Tool Calls
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 2: Streaming with Tool Calls")
    print("=" * 60)

    tool_request = {
        "model": "gemini-2.0-flash",
        "messages": [
            {"role": "system", "content": "You are a helpful assistant. Use tools when needed."},
            {"role": "user", "content": "What is the weather in Paris?"},
        ],
        "tools": [
            {
                "type": "function",
                "function": {
                    "name": "get_weather",
                    "description": "Get the weather for a location",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "location": {"type": "string", "description": "City name"},
                        },
                        "required": ["location"],
                    },
                },
            },
        ],
        "tool_choice": "auto",
        "stream": True,
    }

    print("\nOpenAI-style Request with Tools:")
    print(json.dumps(tool_request, indent=2))

    # Translate to Gemini
    gemini_tool_request = translate_openai_request_to_gemini(tool_request)
    print("\nTranslated to Gemini:")
    print(json.dumps(gemini_tool_request.data, indent=2))

    async with AsyncClient(timeout=Timeout(connect=GEMINI_CONNECT_TIMEOUT_S, read=GEMINI_READ_TIMEOUT_S)) as client:
        async with client.stream(
            "POST",
            GEMINI_STREAM_URL,
            headers={
                "Content-Type": "application/json",
                "x-goog-api-key": GEMINI_API_KEY,
            },
            json=gemini_tool_request.data,
        ) as response:
            if not response.is_success:
                error_text = await response.aread()
                print(f"API Error: {response.status_code} {error_text.decode()}")
            else:
                print("\n--- Streaming Response ---")

                # Collect and display chunks
                lines = []
                async for line in response.aiter_lines():
                    if line.startswith("data: "):
                        json_line = line[6:]
                        if json_line.strip():
                            display = json_line[:100] + ("..." if len(json_line) > 100 else "")
                            print(f"Raw: {display}")
                            lines.append(json_line)

                # Parse and show tool calls
                print("\n--- Parsed Tool Calls ---")
                for line in lines:
                    try:
                        chunk = json.loads(line)
                        parts = chunk.get("candidates", [{}])[0].get("content", {}).get("parts", [])
                        for part in parts:
                            if "functionCall" in part:
                                fc = part["functionCall"]
                                print(f"Tool Call: {fc['name']}({json.dumps(fc.get('args', {}))})")
                    except json.JSONDecodeError:
                        pass

    # =========================================================================
    # Example 3: Stream Aggregation
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 3: Stream Aggregation Demo")
    print("=" * 60)

    # Simulate an OpenAI SSE stream
    simulated_sse_lines = [
        'data: {"id":"chatcmpl-1","choices":[{"delta":{"role":"assistant","content":"Hello"}}]}',
        'data: {"id":"chatcmpl-1","choices":[{"delta":{"content":" world"}}]}',
        'data: {"id":"chatcmpl-1","choices":[{"delta":{"content":"!"}}]}',
        'data: {"id":"chatcmpl-1","choices":[{"delta":{},"finish_reason":"stop"}]}',
        "data: [DONE]",
    ]

    print("\nSimulated OpenAI SSE Stream:")
    for line in simulated_sse_lines:
        print(line)

    # Create async iterator
    async def iter_sse(lines_list):
        for line in lines_list:
            yield line

    # Aggregate the stream
    aggregated = await aggregate_openai_stream(iter_sse(simulated_sse_lines))

    print("\nAggregated Result:")
    print(json.dumps(aggregated.data, indent=2))

    content = aggregated.data.get("choices", [{}])[0].get("delta", {}).get("content")
    print("\nFull Message:", content)

    print("\n" + "=" * 60)
    print("Streaming Examples Complete")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())
